### PR TITLE
Update dockerfile for recently structure changes

### LINKED ISSUE #5245: tag 3.0.2 build container error
```
root@localhost project/tikv (v3.0.2 %) » docker build --network=host -t tikv:v3.0.2_pub -f Dockerfile . 
Sending build context to Docker daemon  32.16MB
Step 1/30 : FROM pingcap/rust as builder
 ---> 215761fcd9f8
Step 2/30 : WORKDIR /tikv
 ---> Using cache
 ---> 6ae0e9b1ad56
Step 3/30 : COPY rust-toolchain ./
 ---> Using cache
 ---> 18f730026bc9
Step 4/30 : RUN rustup default nightly-2019-07-19
 ---> Using cache
 ---> 92b131f513b0
Step 5/30 : COPY Cargo.toml Cargo.lock ./
 ---> Using cache
 ---> aa31dbb33ed5
Step 6/30 : RUN sed -i '/fuzz/d' Cargo.toml &&     sed -i '/test\_/d' Cargo.toml &&     sed -i '/profiler/d' Cargo.toml
 ---> Using cache
 ---> f377d566cd9c
Step 7/30 : COPY Makefile ./
 ---> Using cache
 ---> 1c2d0abb1499
Step 8/30 : COPY scripts/run-cargo.sh ./scripts/run-cargo.sh
 ---> Using cache
 ---> 4cf7afb964c8
Step 9/30 : COPY etc/cargo.config.dist ./etc/cargo.config.dist
 ---> Using cache
 ---> 313053c15966
Step 10/30 : COPY ./components/codec/Cargo.toml ./components/codec/Cargo.toml
 ---> Using cache
 ---> 9604e6d3101d
Step 11/30 : COPY ./components/engine/Cargo.toml ./components/engine/Cargo.toml
 ---> Using cache
 ---> b08bb98ff96e
Step 12/30 : COPY ./components/log_wrappers/Cargo.toml ./components/log_wrappers/Cargo.toml
 ---> Using cache
 ---> 0713765b26f3
Step 13/30 : COPY ./components/match_template/Cargo.toml ./components/match_template/Cargo.toml
 ---> Using cache
 ---> 8d695bb4a264
Step 14/30 : COPY ./components/panic_hook/Cargo.toml ./components/panic_hook/Cargo.toml
 ---> Using cache
 ---> b2953fac4669
Step 15/30 : COPY ./components/tidb_query/Cargo.toml ./components/tidb_query/Cargo.toml
 ---> Using cache
 ---> 7712c9f4ac73
Step 16/30 : COPY ./components/tidb_query_codegen/Cargo.toml ./components/tidb_query_codegen/Cargo.toml
 ---> Using cache
 ---> 49af4112accf
Step 17/30 : COPY ./components/tidb_query_datatype/Cargo.toml ./components/tidb_query_datatype/Cargo.toml
 ---> Using cache
 ---> 163556fecd54
Step 18/30 : COPY ./components/tikv_alloc/Cargo.toml ./components/tikv_alloc/Cargo.toml
 ---> Using cache
 ---> e3e93a0c4a06
Step 19/30 : COPY ./components/tikv_util/Cargo.toml ./components/tikv_util/Cargo.toml
 ---> Using cache
 ---> e6bcf444800a
Step 20/30 : COPY ./components/tipb_helper/Cargo.toml ./components/tipb_helper/Cargo.toml
 ---> Using cache
 ---> c4b2362aeeea
Step 21/30 : RUN mkdir -p ./src/bin &&     echo 'fn main() {}' > ./src/bin/tikv-ctl.rs &&     echo 'fn main() {}' > ./src/bin/tikv-server.rs &&     echo '' > ./src/lib.rs &&     mkdir ./components/codec/src && echo '' > ./components/codec/src/lib.rs &&     mkdir ./components/engine/src && echo '' > ./components/engine/src/lib.rs &&     mkdir ./components/log_wrappers/src && echo '' > ./components/log_wrappers/src/lib.rs &&     mkdir ./components/match_template/src && echo '' > ./components/match_template/src/lib.rs &&     mkdir ./components/panic_hook/src && echo '' > ./components/panic_hook/src/lib.rs &&     mkdir ./components/tidb_query/src && echo '' > ./components/tidb_query/src/lib.rs &&     mkdir ./components/tidb_query_codegen/src && echo '' > ./components/tidb_query_codegen/src/lib.rs &&     mkdir ./components/tidb_query_datatype/src && echo '' > ./components/tidb_query_datatype/src/lib.rs &&     mkdir ./components/tikv_alloc/src && echo '' > ./components/tikv_alloc/src/lib.rs &&     mkdir ./components/tikv_util/src && echo '' > ./components/tikv_util/src/lib.rs &&     mkdir ./components/tipb_helper/src && echo '' > ./components/tipb_helper/src/lib.rs &&     make build_dist_release &&     rm -rf ./target/release/.fingerprint/codec-* &&     rm -rf ./target/release/.fingerprint/engine-* &&     rm -rf ./target/release/.fingerprint/log_wrappers-* &&     rm -rf ./target/release/.fingerprint/match_template-* &&     rm -rf ./target/release/.fingerprint/panic_hook-* &&     rm -rf ./target/release/.fingerprint/tidb_query-* &&     rm -rf ./target/release/.fingerprint/tidb_query_codegen-* &&     rm -rf ./target/release/.fingerprint/tidb_query_datatype-* &&     rm -rf ./target/release/.fingerprint/tikv_alloc-* &&     rm -rf ./target/release/.fingerprint/tikv_util-* &&     rm -rf ./target/release/.fingerprint/tipb_helper-* &&     rm -rf ./target/release/.fingerprint/tikv-*
 ---> Running in a833010eda40
make x-build-dist
make[1]: Entering directory `/tikv'
bash scripts/run-cargo.sh
using etc/cargo.config.dist
+ cp etc/cargo.config.dist .cargo/config
+ set +x
+ cargo build --no-default-features --release -Zunstable-options -Zconfig-profile '--features= jemalloc portable sse'
error: failed to read `/tikv/components/profiler/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
+ r=101
+ set +x
make[1]: *** [x-build-dist] Error 101
make[1]: Leaving directory `/tikv'
make: *** [build_dist_release] Error 2
The command '/bin/sh -c mkdir -p ./src/bin &&     echo 'fn main() {}' > ./src/bin/tikv-ctl.rs &&     echo 'fn main() {}' > ./src/bin/tikv-server.rs &&     echo '' > ./src/lib.rs &&     mkdir ./components/codec/src && echo '' > ./components/codec/src/lib.rs &&     mkdir ./components/engine/src && echo '' > ./components/engine/src/lib.rs &&     mkdir ./components/log_wrappers/src && echo '' > ./components/log_wrappers/src/lib.rs &&     mkdir ./components/match_template/src && echo '' > ./components/match_template/src/lib.rs &&     mkdir ./components/panic_hook/src && echo '' > ./components/panic_hook/src/lib.rs &&     mkdir ./components/tidb_query/src && echo '' > ./components/tidb_query/src/lib.rs &&     mkdir ./components/tidb_query_codegen/src && echo '' > ./components/tidb_query_codegen/src/lib.rs &&     mkdir ./components/tidb_query_datatype/src && echo '' > ./components/tidb_query_datatype/src/lib.rs &&     mkdir ./components/tikv_alloc/src && echo '' > ./components/tikv_alloc/src/lib.rs &&     mkdir ./components/tikv_util/src && echo '' > ./components/tikv_util/src/lib.rs &&     mkdir ./components/tipb_helper/src && echo '' > ./components/tipb_helper/src/lib.rs &&     make build_dist_release &&     rm -rf ./target/release/.fingerprint/codec-* &&     rm -rf ./target/release/.fingerprint/engine-* &&     rm -rf ./target/release/.fingerprint/log_wrappers-* &&     rm -rf ./target/release/.fingerprint/match_template-* &&     rm -rf ./target/release/.fingerprint/panic_hook-* &&     rm -rf ./target/release/.fingerprint/tidb_query-* &&     rm -rf ./target/release/.fingerprint/tidb_query_codegen-* &&     rm -rf ./target/release/.fingerprint/tidb_query_datatype-* &&     rm -rf ./target/release/.fingerprint/tikv_alloc-* &&     rm -rf ./target/release/.fingerprint/tikv_util-* &&     rm -rf ./target/release/.fingerprint/tipb_helper-* &&     rm -rf ./target/release/.fingerprint/tikv-*' returned a non-zero code: 2
```

### PR DESCRIPTION
**Pull request by you06**
## What have you changed? (mandatory)

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
- Summarize your change (mandatory)
Update Dockerfile, fix #5245 
- How does this PR work? Need a brief introduction for the changed logic (optional)
Remove profiler from `Cargo.toml`
Add pd_client component build
- Separately describe each logical change and avoid lazy messages (optional)

- Describe any limitations of the current code (optional)

Please **NOTE** that:
- Do not assume reviewers understand the original issue

## What are the type of changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Bugfix (a change which fixes an issue)

## How has this PR been tested? (mandatory)

Test with command `docker build --network=host -t tikv:v3.0.2_pub -f Dockerfile .` and build docker image success.

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect `tidb-ansible` update? (mandatory)

No

## Refer to a related PR or issue link (optional)

#5245

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -68,6 +78,14 @@ RUN mkdir -p ./src/bin && \
...
 COPY ./src ./src
 COPY ./components ./components
+COPY ./cmd ./cmd
+
+# Remove profiling feature
+RUN sed -i '/^profiling/d' ./cmd/Cargo.toml
+
+# Add cmd back
+RUN sed -i '/"components\/pd_client",/a\ \ "cmd",' Cargo.toml && \
+    sed -i '/X_CARGO_CONFIG_FILE=${DIST_CONFIG}/ax-build-dist: export X_PACKAGE=cmd' Makefile
```
Comment (mapleFU) [inline Dockerfile]: How about just using `COPY` to copy the infomation about cmd back here?

Comment (you06) [reply Dockerfile]: If using `COPY` here, the following commands need to be executed again
```Dockerfile
# Remove fuzz and test workspace, remove profiler feature
RUN sed -i '/fuzz/d' Cargo.toml && \
    sed -i '/test\_/d' Cargo.toml && \
    sed -i '/profiler/d' Cargo.toml
```
I'm not sure which one is better.

**Comment by siddontang** [comment]
please change ./scripts/gen-dockerfile.sh to generate the dockerfile.

**Comment by you06** [comment]
I see this in `./scripts/gen-dockerfile.sh`
```sh
cat <<EOT > ${output}
...
RUN rustup default $(cat "rust-toolchain")
...
EOT
```

Which will generate static toolchain version in Dockerfile

```Dockerfile
RUN rustup default nightly-2019-07-19
```

What about generate Dockerfile like this?

```Dockerfile
RUN rustup default $(cat ./rust-toolchain)
```

**Comment by you06** [comment]
@zhouqiang-cl @mapleFU PTAL again

**Comment by siddontang** [comment]
> RUN rustup default $(cat ./rust-toolchain)

I worry this may evict the docker cache every time, can you approve it?

**Comment by you06** [comment]
> I worry this may evict the docker cache every time, can you check it?

@siddontang  The docker cache works fine, also change rust version in `rust-toolchain` will evict cache. But as there is a dockerfile generator, just regenerate dockerfile after toolchain upgrade is ok.
`RUN rustup default $(cat ./rust-toolchain)` has been removed.

**Comment by Hoverbear** [comment]
Can we make it build with the profiler?

**Comment by you06** [comment]
> Can we make it build with the profiler?

Will the profiler be used in Docker container?

**Pull request merged**